src/walker.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Promise<Pipeline>;
+  stepFn?: (step: Step) => Promise<Step>;
+  commandFn?: (command: Command) => Promise<Command>;
+}
+
+/**
+ * @param p The pipeline to evaluate
+ * @returns An evaluated pipeline with all conditionals generated if they were accepted
+ */
+export async function evaluatePipeline(pipeline: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const newSteps = [];
+  for (const step of pipeline.steps) {
+    const newStep = await evaluateStep(step, conditionalCache);
+    if (newStep) {
+      newSteps.push(newStep);
+    }
+  }
+  pipeline.steps = newSteps;
+  return pipeline;
+}
+
+async function evaluateStep(
+  step: PotentialStep,
+  conditionalCache: Map<any, any>,
+): Promise<Step | null> {
+  if (step instanceof Conditional) {
+    if (await step.accept()) {
+      step = await getAndCacheDependency(conditionalCache, step);
+    } else {
+      return null;
+    }
+  }
+  const deps = step.dependencies;
```

Comment:
This isn't handling effects correctly.

Reply:
I'm not actually sure what effects are,  I read the comment but that doesn't clarify on the purpose behind effects?